RAG_supporters/dataset/label_calculator.py

```diff
@@ -0,0 +1,383 @@
+"""Label calculation system for domain assessment dataset.
+
+Calculates 3 types of labels:
+1. Source/question labels: Based on CSV cluster_probabilities or centroid distances
+2. Steering embedding labels: Based on centroid distances
+3. Combined labels: Average of source and steering labels (for masking augmentation)
+"""
+
+import logging
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import Dict, List, Optional, Tuple, Union
+
+import numpy as np
+from scipy.spatial.distance import cdist
+
+from RAG_supporters.clustering.clustering_data import ClusteringData
+
+logger = logging.getLogger(__name__)
+
+
+class LabelNormalizationMethod(str, Enum):
+    """Supported label normalization methods."""
+    
+    SOFTMAX = "softmax"
+    L1 = "l1"
+    L2 = "l2"
+    SOFTMAX_TEMPERATURE = "softmax_temperature"
+
+
+class LabelNormalizer(ABC):
+    """Base class for label normalization strategies."""
+    
+    @abstractmethod
+    def normalize(self, distances: np.ndarray) -> np.ndarray:
+        """
+        Normalize distances to probability distribution.
+        
+        Args:
+            distances: Array of distances to cluster centroids
+            
+        Returns:
+            Normalized probability vector
+        """
+        pass
+
+
+class SoftmaxNormalizer(LabelNormalizer):
+    """Softmax normalization: exp(-d) / sum(exp(-d))."""
+    
+    def __init__(self, temperature: float = 1.0):
+        """
+        Initialize normalizer.
+        
+        Args:
+            temperature: Temperature parameter for softmax (default 1.0)
+        """
+        self.temperature = temperature
+    
+    def normalize(self, distances: np.ndarray) -> np.ndarray:
+        """Softmax normalization with temperature scaling."""
+        # Use negative distances (closer = higher probability)
+        logits = -distances / self.temperature
+        
+        # Numerical stability: subtract max
+        logits_max = np.max(logits)
+        exp_logits = np.exp(logits - logits_max)
+        
+        # Normalize
+        return exp_logits / np.sum(exp_logits)
+
+
+class L1Normalizer(LabelNormalizer):
+    """L1 normalization: inverse distances normalized to sum to 1."""
+    
+    def normalize(self, distances: np.ndarray) -> np.ndarray:
+        """L1 normalization."""
+        # Add small epsilon to avoid division by zero
+        eps = 1e-10
+        inverse_distances = 1.0 / (distances + eps)
+        return inverse_distances / np.sum(inverse_distances)
+
+
+class L2Normalizer(LabelNormalizer):
```

Comment:
This does NOT produce a probability distribution (doesn't sum to 1). This is likely a bug. Either fix it or remove it.